compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Keyability.hs

```diff
@@ -0,0 +1,127 @@
+    unkeyableTyCons = findUnkeyableTyConsInModule currentModule dataTypes
+
+  for_ (moduleTemplates mod0) $ \tpl ->
+    for_ (tplKey tpl) $ \key ->
+      withContext (ContextTemplate mod0 tpl TPKey) $ do
+        checkKeyType currentModule unkeyableTyCons (tplKeyType key)
+
+findUnkeyableTyConsInModule ::
+     CurrentModule
+  -> NM.NameMap DefDataType
+  -> UnkeyableTyConSet
+findUnkeyableTyConsInModule currentModule dataTypes = do
+  let eqs =
+        [ (dataTypeCon dataType, keyable, deps)
+        | dataType <- NM.toList dataTypes
+        , let (keyable, deps) =
+                case keyabilityConditionsDataType currentModule dataType of
+                  Nothing -> (False, [])
+                  Just deps0 -> (True, HS.toList deps0)
+        ]
+  case leastFixedPointBy (&&) eqs of
+    Left name -> error ("Reference to unknown data type: " ++ show name)
+    Right keyabilities ->
+      UnkeyableTyConSet $ HMS.keysSet $ HMS.filter not keyabilities
+
+-- | Determine whether a data type preserves "keyability".
+keyabilityConditionsDataType ::
+     CurrentModule
+  -> DefDataType
+  -> Maybe (HS.HashSet TypeConName)
+keyabilityConditionsDataType currentModule (DefDataType _loc _ _ _ cons) =
+  mconcatMapM
+    (keyabilityConditionsType currentModule Nothing)
+    (toListOf dataConsType cons)
+
+-- | Determine whether a type is "keyable".
+--
+-- When an 'UnkeyableTyConSet' value is not given, type constructors from the
+-- current module are returned as "keyability" conditions.
+--
+-- When an 'UnkeyableTyConSet' value is given, the presence of a type constructor
+-- from the current module in the set means that it is "unkeyable", and the
+-- returned set is always empty.
+--
+-- Regardless of the 'UnkeyableTyConSet' being provided or not, type constructors
+-- from other modules are assumed to be "keyable".
+keyabilityConditionsType ::
+     CurrentModule
+  -> Maybe UnkeyableTyConSet
+  -> Type
+  -> Maybe (HS.HashSet TypeConName)
+keyabilityConditionsType (CurrentModule currentModuleName) mUnkeyableTyConSet = go
+  where
+    unkeyable = Nothing
+    noConditions = Just HS.empty
+    conditionallyOn = Just . HS.singleton
+
+    go = \case
+      TContractId{} -> unkeyable
+
+      TList typ -> go typ
+      TOptional typ -> go typ
+      TTextMap typ -> go typ
+      TGenMap t1 t2 -> HS.union <$> go t1 <*> go t2
+      TApp tfun targ -> HS.union <$> go tfun <*> go targ
+
+      TCon qtcon
+        | Right tconName <- matching (_PRSelfModule currentModuleName) qtcon ->
+            case mUnkeyableTyConSet of
+              Nothing -> conditionallyOn tconName
+              Just (UnkeyableTyConSet unkeyableTyCons) ->
+                if tconName `HS.member` unkeyableTyCons
+                  then unkeyable
+                  else noConditions
+        | otherwise -> noConditions
+
+      TNumeric{} -> noConditions
+      TNat{} -> noConditions
+      TVar{} -> noConditions
+      TSynApp{} -> noConditions
```

Comment:
Unless you `expandTypeSynonyms` you could get expand to something with a contract id here. This shouldn't happen in damlc tho (iirc type synonyms are only used with constraints, which are always unserializable), so maybe worth adding a comment instead. If we preserved regular type synonyms in LFConversion, this would have to change.

Reply:
my reasoning here was that `DA.Daml.LF.TypeChecker.Serializability` always deems `TSynApp` unserializable without expanding the synonym, so the keyability check doesn't really matter. I'll add a comment to that effect.